proposals/type-classes.md

```diff
@@ -0,0 +1,296 @@
+
+```kotlin
+extension interface Monoid<A> {
+    fun A.combine(b: A): A
+    val empty: A
+}
+```
+
+The above declaration can serve as target for implementations for any arbitrary data type.
+In the implementation below we provide evidence that there is a `Monoid<Int>` instance that enables `combine` and `empty` on `Int`
+
+```kotlin
+package intext
+
+extension object IntMonoid : Monoid<Int> {
+    fun Int.combine(b: Int): Int = this + b
+    val empty: Int = 0
+}
+```
+
+```
+import intext.IntMonoid
+
+1.combine(2) // 3
+IntMonoid.empty() // 0
+```
+
+Because of this constrain where we are stating that there is a `Monoid` constrain for a given type `A` we can also encode polymorphic definitions based on those constrains:
+
+```kotlin
+import intext.IntMonoid
+
+fun <A> add(a: A, b: A, with Monoid<A>): A = a.combine(b)
+add(1, 1) // compiles
+add("a", "b") // does not compile: No `String: Monoid` instance defined in scope
+```
+
+## Compile Time Dependency Injection
+
+On top of the value this brings to typed FP in Kotlin it also helps in OOP contexts where dependencies can be provided at compile time:
+
+```kotlin
+extension interface Context<A> {
+  fun A.config(): Config
+}
+```
+
+```kotlin
+package prod
+
+extension object ProdContext: Context<Service> {
+  fun Service.config(): Config = ProdConfig
+}
+```
+
+```kotlin
+package test
+
+extension object TestContext: Context<Service> {
+  fun Service.config(): Config = TestConfig
+}
+```
+
+```kotlin
+package prod
+
+service.config() // ProdConfig
+```
+
+```kotlin
+package test
+
+service.config() // TestConfig
+```
+
+## Overcoming `inline` + `reified` limitations
+
+Type classes allow us to workaround `inline` `reified` generics and their limitations and express those as type classes instead:
+
+```kotlin
+extension interface Reified<A> {
+    val A.selfClass: KClass<A>
+}
+```
+
+Now a function that was doing something like:
+
+```kotlin
+inline fun <reified A> foo() { .... A::class ... }
+```
+
+can be replaced with:
+
+```kotlin
+fun <A> fooTC(with Reified<A>): Klass<A> { .... A.selfClass ... }
+```
+
+This allows us to obtain generics info without the need to declare the functions `inline` or `reified` overcoming the current limitations of inline reified functions that can't be invoked unless made concrete from non reified contexts.
+
+Not this does not remove the need to use `inline reified` where one tries to instrospect generic type information at runtime with reflection. This particular case is only relevant for those cases where you know the types you want `Reified` ahead of time and you need to access to their class value.
+
+```kotlin
+extension class Foo<A> {
+   val someKlazz = foo<A>() //won't compile because class disallow reified type args.
+}
+```
+
+```kotlin
+extension class Foo<A> {
+   val someKlazz = fooTC<A>() //works and does not requires to be inside an `inline reified` context.
```

Comment:
Why does this work? I don't see any evidence that `Reified<A>` is satisfied here.

Reply:
Wouldn't this require putting `reified` on the classes generic to work properly?